project-koku/koku
# Cost 1407 cost models currency

Fixes [COST-1407](https://issues.redhat.com/browse/COST-1407)

Changes proposed in this PR:
* Adds the ability to pass in currency codes to the cost models for MVP currency support 

Testing Instructions:
1. Pull branch, create and load test customer data
First testing if creating a cost model without a currency will default to USD
2. Hit the following endpoint with the provided data
```
[POST] - http://127.0.0.1:8000/api/cost-management/v1/cost-models/

Data:
{
            "uuid": "source uuid from sources endpoint",
            "name": "test",
            "description": "testing adding model.",
            "source_type" : "OCP",
            "created_timestamp": a Datetime,
            "updated_timestamp": a Datetime,
            "rates": [],
            "markup": {
                "value": "30.0000000000",
                "unit": "percent"
            }
}
```
3.  Hit the following endpoint should have your resulted cost model with currency as USD
```
[GET]-http://127.0.0.1:8000/api/cost-management/v1/cost-models/

{
            "uuid": UUID,
            "name": "test",
            "description": "testing adding model.",
            "source_type": "OpenShift Container Platform",
            "created_timestamp": a Datetime,
            "updated_timestamp": a Datetime,
            "rates": [],
            "markup": {
                "value": "30.0000000000",
                "unit": "percent"
            },
            "distribution": "cpu",
            "currency": "USD",
            "source_uuids": [],
            "sources": []
 }
```
Now update users setting so we can test default to user account settings
4. Hit the following endpoint to update users currency to JPY
```
[POST] - http://127.0.0.1:8000/api/cost-management/v1/settings/

{
    "api": {
        "settings": {
            "currency": "JPY"
        }
    }
}
```
5. Hit the endpoint for creating a cost model again
```
[POST] - http://127.0.0.1:8000/api/cost-management/v1/cost-models/

Data:
{
            "uuid": "source uuid from sources endpoint",
            "name": "test",
            "description": "testing adding model.",
            "source_type" : "OCP",
            "created_timestamp": a Datetime,
            "updated_timestamp": a Datetime,
            "rates": [],
            "markup": {
                "value": "30.0000000000",
                "unit": "percent"
            }
}
```
6. Hit the cost models endpoint to see new created cost model with JPY currency
```
[GET]-http://127.0.0.1:8000/api/cost-management/v1/cost-models/
{
            "uuid": UUID,
            "name": "test",
            "description": "testing adding model.",
            "source_type": "OpenShift Container Platform",
            "created_timestamp": a Datetime,
            "updated_timestamp": a Datetime,
            "rates": [],
            "markup": {
                "value": "30.0000000000",
                "unit": "percent"
            },
            "distribution": "cpu",
            "currency": "JPY",
            "source_uuids": [],
            "sources": []
 }
```
Now testing for passing in a currency code
7. Hit the creating cost model endpoint passing in a currency
```
[POST] - http://127.0.0.1:8000/api/cost-management/v1/cost-models/

Data:
{
            "uuid": "source uuid from sources endpoint",
            "name": "test",
            "description": "testing adding model.",
            "source_type" : "OCP",
            "created_timestamp": a Datetime,
            "updated_timestamp": a Datetime,
            "rates": [],
            "currency": "NOK",
            "markup": {
                "value": "30.0000000000",
                "unit": "percent"
            }
}
```
8.  Hit cost models endpoint again to check if it was created with NOK as currency
```
[GET]-http://127.0.0.1:8000/api/cost-management/v1/cost-models/
{
            "uuid": UUID,
            "name": "test",
            "description": "testing adding model.",
            "source_type": "OpenShift Container Platform",
            "created_timestamp": a Datetime,
            "updated_timestamp": a Datetime,
            "rates": [],
            "markup": {
                "value": "30.0000000000",
                "unit": "percent"
            },
            "distribution": "cpu",
            "currency": "NOK",
            "source_uuids": [],
            "sources": []
 }
```

---
Additional Context:
Successful smoke tests link: https://ci.ext.devshift.net/job/project-koku-koku-pr-check/4496/
